dissect/database/sqlite3/sqlite3.py

```diff
@@ -108,14 +161,39 @@ def indices(self) -> Iterator[Index]:
             yield Index(self, *cell.values)
 
     def raw_page(self, num: int) -> bytes:
+        """Retrieve the raw frame data for the given page number.
+
+        Reads the page from a checkpoint, if this class was initialized with a WAL checkpoint.
+
+        If a WAL is available, will first check if the WAL contains a more recent version of the page,
+        otherwise it will read the page from the database file.
+
+        References:
+            - https://sqlite.org/fileformat2.html#reader_algorithm
+        """
         # Only throw an out of bounds exception if the header contains a page_count.
         # Some old versions of SQLite3 do not set/update the page_count correctly.
         if (num < 1 or num > self.header.page_count) and self.header.page_count > 0:
             raise InvalidPageNumber("Page number exceeds boundaries")
+
         if num == 1:  # Page 1 is root
             self.fh.seek(len(c_sqlite3.header))
-        else:
-            self.fh.seek((num - 1) * self.page_size)
+            return self.fh.read(self.header.page_size)
+        self.fh.seek((num - 1) * self.page_size)
+
+        # If a specific WAL checkpoint was provided, use it instead of the on-disk page.
+        if self.wal and self.checkpoint is not None and num in self.checkpoint:
```

Comment:
```suggestion
        if self.checkpoint is not None and num in self.checkpoint:
```
Can skip the `self.wal` check I think.